docs/proposals/status.md

```diff
@@ -0,0 +1,300 @@
+// +k8s:deepcopy-gen=true
+type ObjectStatus struct {
+	// Link to object
+	Link string `json:"link,omitempty"`
+	// Name of object
+	Name string `json:"name,omitempty"`
+	// Kind of object
+	Kind string `json:"kind,omitempty"`
+	// Object group
+	Group string `json:"group,omitempty"`
+        // Conditions represents the latest state of the object
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions []Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// ExtendedStatus adds Kind specific status information for well known types
+	ExtendedStatus `json:",inline,omitempty"`
+}
+
+// ExtendedStatus is a holder of additional status for well known types
+// +k8s:deepcopy-gen=true
+type ExtendedStatus struct {
+	// StatefulSet status
+	STS *Statefulset `json:"sts,omitempty"`
+	// PDB status
+	PDB *Pdb `json:"pdb,omitempty"`
+}
+
+// Statefulset is a generic status holder for stateful-set
+// +k8s:deepcopy-gen=true
+type Statefulset struct {
+	// Replicas defines the no of MySQL instances desired
+	Replicas int32 `json:"replicas"`
+	// ReadyReplicas defines the no of MySQL instances that are ready
+	ReadyReplicas int32 `json:"readycount"`
+	// CurrentReplicas defines the no of MySQL instances that are created
+	CurrentReplicas int32 `json:"currentcount"`
+	// progress is a fuzzy indicator. Interpret as a percentage (0-100)
+	// eg: for statefulsets, progress = 100*readyreplicas/replicas
+	Progress int32 `json:"progress"`
+}
+
+// Pdb is a generic status holder for pdb
+type Pdb struct {
+	// currentHealthy
+	CurrentHealthy int32 `json:"currenthealthy"`
+	// desiredHealthy
+	DesiredHealthy int32 `json:"desiredhealthy"`
+}
+``` 
+
+### API Status meta
+Meta is an aggregate of the common status fields that should be inlined in the resource status struct.
+
+```go
+package status
+
+// Meta is a generic set of fields for status objects
+// +k8s:deepcopy-gen=true
+type Meta struct {
+	// ObservedGeneration is the most recent generation observed. It corresponds to the
+	// Object's generation, which is updated on mutation by the API Server.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty" protobuf:"varint,1,opt,name=observedGeneration"`
+	// Conditions represents the latest state of the object
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions []Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,10,rep,name=conditions"`
+	// Resources embeds a list of object statuses
+	// +optional
+	ComponentList `json:",inline,omitempty"`
+}
+``` 
+
+### API Application status
+```go
+package application
+// ApplicationStatus defines controllers the observed state of Application
+type ApplicationStatus struct {
+	status.Meta `json:",inline"`	
+}
+```
+
+
+### Example
+```
+status:
+  components:
+  - link: /apis/apps/v1/namespaces/default/statefulsets/basic-di
```

Comment:
this example is not consistent with what described above. Each component should have a list of conditions instead of a single `status` field.

```
status:
  components:
  - link: /apis/apps/v1/namespaces/default/statefulsets/basic-di
    conditions:
    - type: Ready
      status: "True"
    - type: Settled
      status: "True"
```

Reply:
Thanks.
Do you prefer a set of conditions or a singular value ?

Reply:
I prefer set of conditions. My understandings of the proposal are:
- The proposed Ready and Settled types should eventually be universal on all resources.
- They will aggregate up to the Application's Ready and Settled types.
- It will take a while for core controllers to adopt these types.

So the application controller will enable users to derive Ready and Settled conditions of the standard resources now, which is very useful.

Reply:
We will follow up on conditions as an issue